server/src/main/java/io/crate/planner/optimizer/iterative/Memo.java

```diff
@@ -0,0 +1,225 @@
+ * would be stored as:
+ * <pre>
+ * root: G0
+ *
+ * G0 : { A -> G1 }
+ * G1 : { B -> [G2, G3] }
+ * G2 : { C -> G4 }
+ * G3 : { E -> G5 }
+ * G4 : { D }
+ * G5 : { F }
+ * </pre>
+ * Groups are reference-counted, and groups that become unreachable from the root
+ * due to mutations in a subtree get garbage-collected.
+ */
+public class Memo {
+    private static final int ROOT_GROUP_REF = 0;
+
+    private final int rootGroup;
+
+    private final Map<Integer, Group> groups = new HashMap<>();
+
+    private int nextGroupId = ROOT_GROUP_REF + 1;
+
+    public Memo(LogicalPlan plan) {
+        rootGroup = insertRecursive(plan);
+        groups.get(rootGroup).incomingReferences.add(ROOT_GROUP_REF);
+    }
+
+    public int getRootGroup() {
+        return rootGroup;
+    }
+
+    public LogicalPlan resolve(int group) {
+        return group(group).membership;
+    }
+
+    public LogicalPlan resolve(GroupReference groupReference) {
+        return resolve(groupReference.groupId());
+    }
+
+    public LogicalPlan extract() {
+        return extract(resolve(rootGroup));
+    }
+
+    private LogicalPlan extract(LogicalPlan node) {
+        return resolveGroupReferences(node, GroupReferenceResolver.from(this::resolve));
+    }
+
+    private Group group(int group) {
+        if (!groups.containsKey(group)) {
+            throw new IllegalStateException();
+        }
+        return groups.get(group);
+    }
+
+    public LogicalPlan replace(int groupId, LogicalPlan node) {
+        Group group = group(groupId);
+        final LogicalPlan old = group.membership;
+
+        if (node instanceof GroupReference) {
+            node = resolve(((GroupReference) node).groupId());
+        } else {
+            node = insertChildrenAndRewrite(node);
+        }
+
+        incrementReferenceCounts(node, groupId);
+        group.membership = node;
+        decrementReferenceCounts(old, groupId);
+        return node;
+    }
+
+    private void incrementReferenceCounts(LogicalPlan fromNode, int fromGroup) {
+        Set<Integer> references = allReferences(fromNode);
+        for (int group : references) {
+            groups.get(group).incomingReferences.add(fromGroup);
+        }
+    }
+
+    private void decrementReferenceCounts(LogicalPlan fromNode, Integer fromGroup) {
+        Set<Integer> references = allReferences(fromNode);
+
+        for (int group : references) {
+            Group childGroup = groups.get(group);
+            if (!childGroup.incomingReferences.remove(fromGroup)) {
+                throw new IllegalStateException("Reference to remove not found");
+            }
+            ;
+            if (childGroup.incomingReferences.isEmpty()) {
+                deleteGroup(group);
+            }
+        }
+    }
+
+    private Set<Integer> allReferences(LogicalPlan node) {
+        return node.sources().stream()
+            .map(GroupReference.class::cast)
+            .map(GroupReference::groupId)
+            .collect(Collectors.toSet());
```

Comment:
Can there really be duplicated? Otherwise the increments/decrements  logic could iterate over the sources without allocating a new set.

Reply:
This is taken from the original implementation, not 100 % sure. I think a plan can point to multiple groups. I will keep that in mind when progressing with the implementation.